palantir-java-format-spi/src/main/java/com/palantir/javaformat/java/FormatterService.java

```diff
@@ -48,4 +48,15 @@ ImmutableList<Replacement> getFormatReplacements(String input, Collection<Range<
      *     Style Guide - 3.3.3 Import ordering and spacing</a>
      */
     String formatSourceReflowStringsAndFixImports(String input) throws FormatterException;
+
+    /**
+     * Fixes imports (eg. ordering, spacing, and removal of unused import statements).
+     *
+     * @param input the input string
+     * @return the output string
+     * @throws FormatterException if the input string cannot be parsed
+     * @see <a href="https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing">Google Java
+     *     Style Guide - 3.3.3 Import ordering and spacing</a>
+     */
+    String fixImports(String input) throws FormatterException;
```

Comment:
I think this will cause issues when there is an old version of the Gradle plugin but this new version of IntelliJ plugin exists. The old `BootstrappingFormatterService` will be service loaded and will not have this method, which the JVM will not be happy about.

Making this a default method that returns `Optional<>` would fix this, so we could detect if the `fixImports` method doesn't exist.

However, we also have the opposite problem - when the version of the Gradle plugin is new but the IntelliJ plugin is old, we'll try to load a `BootstrappingFormatterService` that has an extra method that does not exist in the interface - I'm not 100% sure whether this is a break, but I think the JVM is going to be unhappy again. Needs some real testing.

I'm concerned this might have been built in a way that means we can't really change it in a way that's both forwards and backwards compatible.

Reply:
The solution is probably to make another interface and "make do" with the previous one - ie formatting and import ordering at the same time possibly with a warning.

Reply:
I say all this, but testing it out it seems to work somehow in `./gradlew runIde` with older versions of the Gradle plugin - maybe it's fine or there's something I don't understand happening with the classloaders here. I'll give the other way around (old intellij plugin, new gradle plugin) a manual test too.

Reply:
Newer gradle plugin, older intellij plugin also seems to work 🤔

Reply:
> I think this will cause issues when there is an old version of the Gradle plugin but this new version of IntelliJ plugin exists. The old BootstrappingFormatterService will be service loaded and will not have this method, which the JVM will not be happy about.

When I was running the tests using `./gradlew runIde` the newest version of Intelij was running (2024.1). I opened a project that had an old `gradle-palantir-java-format` plugin applied and the formatter worked. The formatter that was picked up was bundled from the palantir-java-plugin. There seems to be this logic [here](https://github.com/palantir/palantir-java-format/blob/2ec339a038e24c78b27509af9f2c7cd20ac798d5/idea-plugin/src/main/java/com/palantir/javaformat/intellij/FormatterProvider.java#L114) which gets set up [here](https://github.com/palantir/palantir-java-format/blob/2ec339a038e24c78b27509af9f2c7cd20ac798d5/idea-plugin/src/main/java/com/palantir/javaformat/intellij/PalantirJavaFormatSettings.java#L91). So from what I understand, the higher version of the formatter will always be picked up. 

> However, we also have the opposite problem - when the version of the Gradle plugin is new but the IntelliJ plugin is old, we'll try to load a BootstrappingFormatterService that has an extra method that does not exist in the interface - I'm not 100% sure whether this is a break, but I think the JVM is going to be unhappy again. Needs some real testing.

Tested the workflow by bumping my Intelij version to latest. (2024.1 - without my fix) and used the locally published `gradle-palantir-java-format` (with the fix). The formatter worked correctly (and I didn't see any errors in the idea.log file)

> The solution is probably to make another interface and "make do" with the previous one - ie formatting and import ordering at the same time possibly with a warning.
Wdyt if I do create a new interface that will only be used by the Intelij plugin ?